src/backend.f90

```diff
@@ -0,0 +1,174 @@
+      procedure(transposer), deferred :: trans_x2y, trans_x2z
+      procedure(sum9into3), deferred :: sum_yzintox
+   end type base_backend_t
+
+   abstract interface
+      subroutine transeq_ders(self, du, duu, d2u, u, v, w, conv, derps)
+         implicit none
+         import :: base_backend_t
+         import :: field_t
+
+         class(base_backend_t) :: self
+         class(field_t), intent(out) :: du, duu, d2u
+         class(field_t), intent(in) :: u, v, w, conv
+         type(derps_t), intent(in) :: derps
+      end subroutine transeq_ders
+   end interface
+
+   abstract interface
+      subroutine transposer(self, u_, v_, w_, u, v, w)
+         implicit none
+         import :: base_backend_t
+         import :: field_t
+
+         class(base_backend_t) :: self
+         class(field_t), intent(out) :: u_, v_, w_
+         class(field_t), intent(in) :: u, v, w
+      end subroutine transposer
+   end interface
+
+   abstract interface
+      subroutine sum9into3(self, du, dv, dw, du_y, dv_y, dw_y, du_z, dv_z, dw_z)
+         import :: base_backend_t
+         import :: field_t
+
+         class(base_backend_t) :: self
+         class(field_t), intent(inout) :: du, dv, dw
+         class(field_t), intent(in) :: du_y, dv_y, dw_y, du_z, dv_z, dw_z
+      end subroutine sum9into3
+   end interface
+
+contains
+
+   subroutine transeq(self, du, dv, dw, u, v, w)
+      implicit none
+
+      class(base_backend_t) :: self
+      class(field_t), intent(out) :: du, dv, dw
+      class(field_t), intent(in) :: u, v, w
+
+      class(field_t), pointer :: u_y, v_y, w_y, u_z, v_z, w_z, du_y, dv_y, dw_y, du_z, dv_z, dw_z
+
+      !1/2(nabla u curl u + u nabla u) + nu nablasq u
+
+      ! call derivatives in x direction. Based on the run time arguments this
+      ! executes a distributed algorithm or the Thomas algorithm.
+      call transeq_x(du, dv, dw, u, v, w, u, self%xderps)
+
+      ! request fields from the allocator
+      u_y => self%allocator%get_block()
+      v_y => self%allocator%get_block()
+      w_y => self%allocator%get_block()
+      du_y => self%allocator%get_block()
+      dv_y => self%allocator%get_block()
+      dw_y => self%allocator%get_block()
+
+      ! reorder data from x orientation to y orientation
+      call trans_x2y(u_y, v_y, w_y, u, v, w)
+      ! similar to the x direction, obtain derivatives in y.
+      call transeq_y(du_y, dv_y, dw_y, u_y, v_y, w_y, v_y, self%yderps)
+
+      ! we don't need the velocities in y orientation any more, so release
+      ! them to open up space.
+      ! It is important that this doesn't actually deallocate any memory,
+      ! it just makes the corresponding memory space available for use.
+      call self%allocator%release_block(u_y)
+      call self%allocator%release_block(v_y)
+      call self%allocator%release_block(w_y)
+
+      ! just like in y direction, get some fields for the z derivatives.
+      u_z => self%allocator%get_block()
```

Comment:
Yes when we're doing the pointer assignment we actually need to specify bounds so that it matches the orientation we want. Its called bounds remapping and was introduced with the Fortran 2003 standard. I tested this with nvfortran and gfortran, haven't got any issues.

Reply:
In this case `u_y` is only a memory block and there is no need for a specific class ?

Reply:
Well actually it has to be u_y%data(1:SZ, 1:n_y, 1:n_ygroups) => ...

Reply:
Hmm actually we won't be able to do this. The data component is different for cuda and cpu backends. So here we just associate a field without any bounds, its just a pointer after all. Then in backends we use a pointer to a regular array or device array and point the data component with defining the bounds.

Reply:
Ok. Then the get_block function probably needs some arguments.

Reply:
x/y/zderps has got the dimensionality info and they're passed to the transeq_? subroutines so I think it is better to arrange it via derps. Otherwise the grid info is like duplicated both in field type and derps type and its extra work to make sure they're the same.

Reply:
Avoiding duplication makes sense. However, fields will be used for derivations / interpolations but also for other operations, such as IO. Thus, it could be relevant to include the grid/dimensionality info inside the field.

Without providing arguments to the subroutine `get_block`, I think the pointer u_y%data will not have the correct shape/bounds by default, at least right after the call to get_block. This could be an issue.

Reply:
https://henriquesd.medium.com/dry-kiss-yagni-principles-1ce09d9c601f

Reply:
You're trying to violate YAGNI and DRY :smiley:

Reply:
The size of the field is actually determined when we instantiate the allocator and is always the same size. And because backend needs to access allocator, allocator needs to be instantiated first.
https://github.com/xcompact3d/x3d2/blob/main/src/allocator.f90#L71
https://github.com/xcompact3d/x3d2/blob/main/tests/test_allocator.f90#L14

Each time we request a field exactly predetermined size of memory is provided to us. The only problem is when we switch pencil orientation the bounds differ a little for different dimensions.
x pencil data is (SZ, n_x, n_x_groups) and y pencil is (SZ, n_y, n_y_groups). And to support this we'll probably need to pad n_x, n_y, and n_z in the global grid sizes. So most of the time we'll have unneccessary regions in the field array. 
And this unnecessary region depends on wheter we're storing velocity or pressure, but we always have a way to access this information. If it is a u_y and then it is a velocity array in y orientation, and we can get this via yderps type. We need the n_y and n_y_groups even for executing a GPU kernel for example. If we try to obtain these sizes from the field type then it is not the straightforward way and more error prone in my opinion. There are many fields passed to transeq_y for example, and obtaining the grid size from one lucky field is not the best. This grid size is very important to transeq_y, and we pass this info already when we call transeq_y via yderps type.